File: Wayfork/WayforkCore/Sources/WayforkCore/Support/IPv4Prefix.swift
import Foundation

/// An IPv4 address or CIDR prefix: the parsed form of an IP rule pattern (F11) and the
/// arithmetic behind the auto-route exclusion list (`SingBoxConfigGenerator`).
public struct IPv4Prefix: Hashable, Sendable, CustomStringConvertible {
    public let address: UInt32
    public let bits: Int

    /// Host bits are cleared: `10.8.0.5/24` becomes `10.8.0.0/24`.
    public init(address: UInt32, bits: Int) {
        self.bits = bits
        self.address = address & Self.mask(bits)
    }

    /// Parses `a.b.c.d` (a /32) or `a.b.c.d/n`; host bits are cleared. Dotted-quad IPv4
    /// only — no IPv6, no shorthand like `10/8`.
    public init?(_ text: String) {
        let parts = text.split(separator: "/", maxSplits: 1, omittingEmptySubsequences: false)
        var bits = 32
        if parts.count == 2 {
            guard parts[1].allSatisfy(\.isNumber), let n = Int(parts[1]), (0...32).contains(n)
            else { return nil }
            bits = n
        }
        var raw = in_addr()
        guard String(parts[0]).withCString({ inet_pton(AF_INET, $0, &raw) }) == 1 else {
            return nil
        }
        self.init(address: UInt32(bigEndian: raw.s_addr), bits: bits)
    }

    static func mask(_ bits: Int) -> UInt32 {
        bits == 0 ? 0 : ~UInt32(0) << UInt32(32 - bits)
    }

    /// A single address.
    public var isHost: Bool { bits == 32 }

    public func contains(_ other: IPv4Prefix) -> Bool {
        other.bits >= bits && (other.address & Self.mask(bits)) == address
    }

    /// Prefixes either nest or are disjoint, so overlapping means one contains the other.
    public func overlaps(_ other: IPv4Prefix) -> Bool {
        contains(other) || other.contains(self)
    }

    /// `self` minus `inner`, as prefixes: the sibling of every step down from `self` to
    /// `inner`, sorted by address. Returns `[self]` when `inner` lies outside, `[]` when
    /// `inner` covers `self`.
    public func subtracting(_ inner: IPv4Prefix) -> [IPv4Prefix] {
        guard contains(inner) else { return [self] }
        guard inner.bits > bits else { return [] }
        return ((bits + 1)...inner.bits)
            .map { depth in
                IPv4Prefix(address: inner.address ^ (UInt32(1) << UInt32(32 - depth)), bits: depth)
            }
            .sorted { $0.address < $1.address }
    }

    /// `self` minus every prefix in `inners`, sorted by address.
    public func subtracting(all inners: [IPv4Prefix]) -> [IPv4Prefix] {
        var result = [self]
        for inner in inners {
            result = result.flatMap { $0.subtracting(inner) }
        }
        return result.sorted { $0.address < $1.address }
    }

    /// `a.b.c.d/n`, always with the prefix length.
    public var description: String { "\(dotted)/\(bits)" }

    /// The stored pattern form of an IP rule: a bare address for a /32, `a.b.c.d/n` otherwise.
    public var canonical: String { isHost ? dotted : description }

    private var dotted: String {
        "\(address >> 24).\((address >> 16) & 0xff).\((address >> 8) & 0xff).\(address & 0xff)"
    }
}

File: WayforkWindows/app/lib/core/support/ipv4_prefix.dart
final class IPv4Prefix {
  IPv4Prefix(int address, int bits)
    : bits = RangeError.checkValueInInterval(bits, 0, 32, 'bits'),
      address = address & _mask(bits);

  final int address;
  final int bits;

  static IPv4Prefix? parse(String text) {
    final slash = text.indexOf('/');
    if (slash != -1 && text.indexOf('/', slash + 1) != -1) return null;
    final addressText = slash == -1 ? text : text.substring(0, slash);
    final bitsText = slash == -1 ? null : text.substring(slash + 1);
    if (bitsText != null &&
        (!_plainDigits(bitsText) || int.parse(bitsText) > 32)) {
      return null;
    }
    final parts = addressText.split('.');
    if (parts.length != 4) return null;
    var address = 0;
    for (final part in parts) {
      if (!_plainDigits(part) ||
          part.length > 3 ||
          (part.length > 1 && part.startsWith('0'))) {
        return null;
      }
      final octet = int.parse(part);
      if (octet > 255) return null;
      address = (address << 8) | octet;
    }
    return IPv4Prefix(address, bitsText == null ? 32 : int.parse(bitsText));
  }

  static int _mask(int bits) =>
      bits == 0 ? 0 : (0xffffffff << (32 - bits)) & 0xffffffff;

  bool get isHost => bits == 32;

  bool contains(IPv4Prefix other) =>
      other.bits >= bits && (other.address & _mask(bits)) == address;

  bool overlaps(IPv4Prefix other) => contains(other) || other.contains(this);

  List<IPv4Prefix> subtracting(IPv4Prefix inner) {
    if (!contains(inner)) return [this];
    if (inner.bits <= bits) return const [];
    final result = <IPv4Prefix>[];
    for (var depth = bits + 1; depth <= inner.bits; depth++) {
      result.add(IPv4Prefix(inner.address ^ (1 << (32 - depth)), depth));
    }
    result.sort((a, b) => a.address.compareTo(b.address));
    return result;
  }

  List<IPv4Prefix> subtractingAll(Iterable<IPv4Prefix> inners) {
    var result = <IPv4Prefix>[this];
    for (final inner in inners) {
      result = [for (final prefix in result) ...prefix.subtracting(inner)];
    }
    result.sort((a, b) => a.address.compareTo(b.address));
    return result;
  }

  String get _dotted =>
      '${address >> 24}.${(address >> 16) & 0xff}.${(address >> 8) & 0xff}.${address & 0xff}';

  String get canonical => isHost ? _dotted : toString();

  @override
  String toString() => '$_dotted/$bits';

  @override
  bool operator ==(Object other) =>
      other is IPv4Prefix && address == other.address && bits == other.bits;

  @override
  int get hashCode => Object.hash(address, bits);
}

bool _plainDigits(String value) {
  if (value.isEmpty) return false;
  for (final code in value.codeUnits) {
    if (code < 0x30 || code > 0x39) return false;
  }
  return true;
}
